Add spectrum support

WTFox opened the pull request:
Hello! :wave: 🙂 We are picking up where #243 left off.

This pull request's main goal was to pass the tests without changing any of @mtami's and @briandailey's excellent work. My changes consisted of only formatting and removing python 3.5 in tests. Initially, I wanted to recreate the previous failure, but since python 3.5 is end-of-life, I couldn't even get it installed on my 2019 MBP running Ventura. I created https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/262 to decide which python versions we wish to maintain support for.

Tests are passing locally.
<img width="552" alt="image" src="https://user-images.githubusercontent.com/2353242/216710338-9aca6625-aa82-4819-b4a7-ba00eccd6bd0.png">

Please let me know if I've missed anything!

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

WTFox commented:
I updated CHANGES.rst to reflect that we no longer support Python 3.5 and SQLAlchemy 1.3. If that's okay, I need to probably update setup.py, too. If so, could you provide the values you'd like for the following attributes:

- `install_requires`
- `python_requires`
- `classifiers`

jasonamyers commented:
are these deprecations enough that there should be a major version bump?

Brooke-white commented:
@jasonamyers, thats a good question. I think a major version bump would be needed for dropping sqlalchemy 1.3 support..though I am unsure how other libraries have handled dropping support for unsupported versions of python. Any idea?

For some comparison to [other sqlalchemy dialects](https://docs.sqlalchemy.org/en/20/dialects/), I browsed a few and saw many still support sqlalchemy 1.3. 

@WTFox , Do we know if this PR explicitly breaks sqlalchemy 1.3 support? I'd be interested to see how this PR tests for sqlalchemy 1.3. Per [sqlalchemy's 1.3.23 release](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_23/setup.py#L178) it should support python 3.4-3.9, so let's try running against a newer not-unsupported Python version and see what results we get.

If we do see breaking tests, unsure how much effort we want to place into supporting a legacy version of sqlalchemy if the fixes are non-trivial. On a related note, sqlalchemy 2.0 was recently released, and I will open an issue for supporting that in the future. But it is for another day :)

tagging @graingert here to get his opinion on this.

WTFox commented:
@Brooke-white Good call! I reinstated tests to test against SQLAlchemy version 1.3 using python 3.9. I also updated the PR body and CHANGES.rst to reflect this.

jasonamyers commented:
@WTFox how would you feel about adding tox tests for 3.10 for the higher rev sqlalchemy versions?

WTFox commented:
@jasonamyers I added tests for the following:

- Python 3.10 - SQLA 1.4
- Python 3.10 - SQLA 2.0

This actually surfaced some changes in pytest parametrized tests. I updated the tests in test_reflection.py to utilize the newer `pytest.param` function. See [this StackOverflow answer for context(https://stackoverflow.com/a/30575822).

WTFox commented:
Quick update:

After adding SQLAlchemy 2.0 tests, a few issues surfaced that I'm working through now.

Brooke-white commented on sqlalchemy_redshift/dialect.py:
```diff
@@ -625,6 +620,84 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+                r"^CHECK *\((.+)\)( NOT VALID)?$", src, flags=re.DOTALL
+            )
+            if not m:
+                print(f"Could not parse CHECK constraint text: {src}")
```
do we want to print or log here? idk if we have existing log statements in the project, but it could be good to introduce one in this case

WTFox replied:
Logging would be better, I can change this.

Brooke-white reviewed (commented):
this may be out of scope for this PR...but figured i'd mention as you're working in these methods incase the effort to do this is minimal:

when executing queries, could we use bind variables  in-place of string concatenation? we'd want to have some logic to set the paramstyle to some common type (e.g. format), then re-set to the original value post query execution. only caveat with this is we cant use bind variables for sub-queries (i dont think i saw any but there may be some), due to limitations on Redshift server.

I know we're not using bind parameters in the existing code base, but we _really_ should be  as it provides an added level of protection against sql injection. if it's too much extra effort for now just let me know and I can open an issue for addressing this in the future :)

WTFox commented:
This PR is ready for review. We decided against adding SQLAlchemy 2.0 compatibility. There is a private SQLAlchemy dialect function that we're using, `_get_column_info` that no longer exists in the dialect, and the effort required to fix that exceeds the scope of this work. See [this](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#major-architectural-performance-and-api-enhancements-for-database-reflection).

Tests are passing for the following tox envs:
<img width="763" alt="image" src="https://user-images.githubusercontent.com/2353242/217374618-585662bd-b520-4b25-b2f2-34e28bb740d4.png">

A few noteworthy items:
- docs were failing, but pinning sphinx to 1.6 fixed it
- support for python 3.11 was added in psycopg2 2.9, so we're using that for the python 3.11 tests
- alembic was pinned at a higher version to fix some errors as well.

WTFox commented:
> if it's too much extra effort for now just let me know and I can open an issue for addressing this in the future :)

If it works for you, I'd prefer to open a new issue so that this PR can move forward. :)

Brooke-white commented:
sounds good -- opened #265 wrt the bind parameter changes. As our CI is broken right now, I'll kick off these tests on my end and confirm everything looks good. I should get back to you before end of week with results. thanks again for taking on adding spectrum support! :)

WTFox commented:
> thanks again for taking on adding spectrum support! :)

I'm just standing on the shoulders of @jasonamyers @briandailey and @mtami :)

Brooke-white commented:
Hey @WTFox -- I ran the CI test suite on my local machines and encountered a couple failures...it looks like a bunch since we run with so many configurations, but with a quick look I just saw 3 tests failing.

1. `tests/test_reflection.py::test_external_table_reflection`
2. `tests/test_reflection.py::test_reflection`
3. `tests/test_long_tablename.py::test_long_tablename`

Some quick analysis on each...
1. we try to execute some sql using role `'arn:aws:iam::123456789012:role/mySpectrumRole'`. this execution fails because the role does not exist in the account where my test cluster is. but it will not exist in any account, since it's an example role :)...so we will need to create an IAM role, which isn't an issue...but I would like to avoid exposing the role ARN (as it includes the AWS account number), so if we could use some means similar to what we use for `PGPASSWORD` e.g. env var to pass this value into the test framework to dynamically pass to the test framework, I think we could accomplish this. very much open to other ideas though :)
2. The expected sql isn't generated. I didn't look at this one too much, but it appears some clauses may be missing.
3. hits some "database is being accessed by others"


for 3, I remember this error from the prior PR. you can read [the convo there](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243#issuecomment-1027109938) for some context. Prior analysis by mtami@ pointed towards transaction isolation level as a potential culprit. 

I've attached output from my run of the tests via `tox` below -- I know this is far from ideal as we should really have our CI up and running, but a means to and end until then :)

For debugging these I'd recommend to checkout my convo with mtami@ in the previous PR and to consider using a Redshift trial period for some lightweight debugging of the few tests for which issue is seen. 


[spectrum-support-tox.txt](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/files/10690154/spectrum-support-tox.txt)

WTFox commented:
@Brooke-white Thanks for the follow-up! Next steps for me will be to set up a redshift instance and see if I can reproduce those integration test failures locally. That might take a bit, but I'll try to keep you updated along the way.

WTFox commented:
Happy Friday!

With the help of the wonderful @jasonamyers, I got this test suite up and running against our redshift instance. This was made possible by adding a few pytest fixtures with default values but allowing environment variables to override them. Here's a list of environment variables supported:

- PGPASSWORD
- REDSHIFT_HOST
- REDSHIFT_PORT
- REDSHIFT_USERNAME
- REDSHIFT_DATABASE
- AWS_ACCOUNT_ID
- REDSHIFT_IAM_ROLE_NAME

IAM role ARNs are constructed using these values. I've written quite a bit for each new fixture in conftest.py, but putting these in the readme or a contributing doc might be a good idea. Where would be the best place for documentation on running the tests?

We also fixed a bug in the reflection query that _might_ have been the cause of the "DB is being accessed by others" issue. The entire test suite passes when I specify a single tox environment, but I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Finally, I removed the environments for python 3.11 and psycopg 2.9 as this would add extra work to the scope of this PR.

Please take a look and let me know if anything else stands out. Thank you for the help! 🙂

Brooke-white commented:
Hi @WTFox -- apologies for the delay. Taking a look at this PR now.

Brooke-white commented on tests/conftest.py:
```diff
@@ -19,10 +22,137 @@
+
+@pytest.fixture(scope="session")
+def iam_role_name():
+    """ The iam_role_name fixture allows the developer to pass in their own
```
could we add some details around how this role should be configured? e.g. what permissions does it grant? is it associated with the cluster?

WTFox replied:
Hmm. Unfortunately, since we were testing this, we gave the role all permissions to Redshift. However, we did have to associate the role to the cluster so I'll update this docstring and the one below to reflect this.

Brooke-white commented on tests/conftest.py:
```diff
@@ -19,10 +22,137 @@
+@pytest.fixture(scope="session")
+def iam_role_arns():
+    """
+    The iam_role_arns fixture allows the developer to pass in their own
```
similar comment to above -- what permissions do the roles have, when do you want to define this env var, are these roles associated with the cluster?

Brooke-white commented:
> I've written quite a bit for each new fixture in conftest.py, but putting these in the readme or a contributing doc might be a good idea. Where would be the best place for documentation on running the tests?

Thank you for making all of these improvements to our test suite! I think a section in the read me would be helpful. I added some comments wrt the role(s) that could be addressed in the readme. also, if there is any pre-test setup (e.g. table creation) required on the cluster, or region restriction that would be useful details also.

> The entire test suite passes when I specify a single tox environment, but I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Working on this now...our CI is still broken, so I'm working on setting up something temporary in my aws account so tests can at least be kicked off via a remote machine. In parallel, I'm running these tests on my local machine.

Brooke-white reviewed (commented):
overall look good -- i will add some more comments once I wrap up getting the test suite kicked off locally + on a remote machine

Brooke-white commented on tests/conftest.py:
```diff
@@ -19,10 +22,137 @@
+
+
+@pytest.fixture(scope="session")
+def iam_role_name():
```
for this logic around the role name & constructing the role arn from various pieces, could we rework this to use look for an env var e.g. `REDSHIFT_IAM_ROLE_ARN` that is the whole arn e.g. `arn:aws:iam::12345678912:role/ExampleRole`

WTFox replied:
This was actually by choice. Some tests look for the ARN, and others look for the account_id and role_name. Passing in an ARN that doesn't reflect the account_id and role_name causes a few test failures due to the mismatch. 

The alternative would be to pass in the role arn and have the account_id, and role name fixtures parse that to get their values. If that's preferred, I can change my implementation.

Brooke-white replied:
Sounds good, yes please update when you get a chance :)

Brooke-white commented:
>  I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Was this issue seen using `tox --parallel`? At least so far, I haven't been able to repo it using a very much "built in a half-day" CI setup 😄 . Maybe the OOM is specific to some limits hit on your machine that are not hit on the ec2 instance used in the CI? I'll keep trying to repro...

WTFox commented:
> Was this issue seen using `tox --parallel`?

Yeah, I was getting a successful test run when I'd specify a single env. I'm almost 100% certain this is due to my personal computer constraints.

Brooke-white commented:
@WTFox -- I'm considering moving our CI test execution to non-parallelized, due to hitting different sorts of errors when run in parallel that are specifically related to the tests running in parallel i.e. external schema already exists because it was created in another test session running at the same time. I don't want to beat these parallelized test issues into the ground, since it has more to do with the test framework than your change, and want to go ahead and get your changes merged once the doc updates + changes around ARN are made :) 

fwiw i believe running tests in parallel was a cost based decision as the previous maintainer was using a personal AWS account. given we don't run this CI too often, i think it should be alright to move to non-parallelized tests at least in the short term.

I'll do a final pass through our CI once the mentioned changes are pushed and then expect to get this merged!

WTFox commented:
Sounds good, I'll work on this tomorrow. Thank you for working through this with me!

Brooke-white commented:
taking a look -- will run through CI

WTFox commented:
> taking a look -- will run through CI

I have a local test run that's almost done, it should be done any minute now. It might be best to hold off until I get that result. I can tag you when ready. Sorry for the lack of communication there. Updating the README is still on my todo, too. But I can do that while CI is running.

Brooke-white commented:
👍 sounds good -- I'll hold off until all else is pushed

WTFox commented:
@Brooke-white, my test run passed. Kick off the CI run at your convenience, and I'll start updating the README.

Brooke-white commented:
Thank you @WTFox @jasonamyers @mtami and all else who helped with this change! I am unavailable next week, but anticipate making a release before end of this month.

matthewgdv commented:
Hi all, just wanted to mention a regression added in the PR that has forced me to pin `sqlalchemy-redshift==0.8.12` in my organization's library code repo. This happens specifically when using a `future=True` engine in sqlalchemy1.4, which previously worked fine. This is not the same as using sqlalchemy2.0, which has made large code changes to its connection and reflection logic relative to 1.4 (even with `FutureEngine`) that break `sqlalchemy-redshift` in other ways.

Basically, this PR added usages of `connection.connect()`  in the reflection code without checking if the connection was already open, which against a `FutureEngine` causes an error because connection forking is no longer allowed from sqlalchemy2.0 onwards.

I can open a new issue if that's better for tracking purposes?

Brooke-white commented:
Hi @matthewgdv , yes please open a new issue so we can work to fix this
